book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -165,49 +220,232 @@ each rooted at the highest finalized block. Each chain consists of a map from
+them to become available.
+
+`ChainState` is defined by the following structure and API:
+
+```rust
+struct ChainSet {
+    chains: BTreeSet<Chain>,
+
+    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
+    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
+    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
+}
+```
+
+#### `pub fn finalize(&mut self) -> Arc<Block>`
+
+Finalize the lowest height block in the non-finalized portion of the best
+chain and updates all side chains to match.
+
+1. Extract the best chain from `self.chains` into `best_chain`
+
+1. Extract the rest of the chains into a `side_chains` temporary variable, so
+   they can be mutated
+
+1. Remove the lowest height block from the best chain with
+   `let block = best_chain.pop_root();`
+
+1. Add `best_chain` back to `self.chains`
+
+1. For each remaining `chain` in `side_chains`
+    - If `chain` starts with `block`, remove `block` and add `chain` back to
+    `self.chains`
+    - Else, drop `chain`
+
+1. for each `height` in `self.queued_by_height` where the height is lower than the
+   new reorg limit
+   - for each `hash` in `self.queued_by_height` at `height`
+     - Remove the key `hash` from `self.queued_blocks` and store the removed `block`
+     - Find and remove `hash` from `self.queued_by_parent` using `block.parent`'s hash
+
+1. Return `block`
+
+### `pub fn queue(&mut self, block: QueuedBlock)`
+
+Queue a non-finalized block to be committed to the state.
+
+After queueing a non-finalized block, this method checks whether the newly
+queued block (and any of its descendants) can be committed to the state
+
+1. Check if the parent block exists in any current chain
+
+1. If it does, call `let ret = self.commit_block(block)`
+    - Call `self.process_queued(new_parents)` if `ret` is `Some`
+
+1. Else Add `block` to `self.queued_blocks` and related members and return
+
+### `fn process_queued(&mut self, new_parent: block::Hash)`
+
+1. Create a list of `new_parents` and populate it with `new_parent`
+
+1. While let Some(parent) = new_parents.pop()
+    - for each `hash` in `self.queued_by_parent.remove(&parent.hash)`
+      - lookup the `block` for `hash`
+      - remove `block` from `self.queued_blocks`
+      - remove `hash` from `self.queued_by_height`
+      - let result = `self.commit_block(block)`;
+      - add `result` to `new_parents`
+
+### `fn commit_block(&mut self, block: QueuedBlock) -> Option<block::Hash>`
+
+Try to commit `block` to the non-finalized state. Returns `None` if the block
+cannot be committed due to missing context.
+
+1. For each `chain`
+    - if `block.parent` == `chain.tip`
+      - try to push `block` onto that chain
+      - broadcast `result` via `block.rsp_tx`
+      - return Some(block.hash) if `result.is_ok()`
+
+1. Find the first chain that contains `block.parent` and fork it with
+  `block.parent` as the new tip
+    - `let fork = self.chains.iter().find_map(|chain| chain.fork(block.parent));`
```

Comment:
It might actually be cheaper to find the chain that contains block.parent at the smallest depth, which will typically be the shortest chain that contains block.parent.

But in practice, I don't think this will matter.